### PR TITLE
Added crude distillation

### PR DESCRIPTION
**Pull request by junmuthy**
I added a `Distil` lattice surgery primitive, as well as corresponding cost functions.  The cost functions are computed using the resource estimator itself on the distillation circuit, and the distil cost function uses the architecture to compute the cost.  This can be improved later by hand after specific analysis for each architecture.  **Currently only T state preparation is supported**.

I have also added a `Distillery` layout which is a derivative of the `Sandwich` layout but now the T factories have ancillas that mirror the usual 15-to-1 BK distillation setup.  Right now those ancillas are isolated in the layout because it works to do it, but later the layout can be expanded to include more highway ancillas.  The main way the distillation procedure enters the process right now is through the `teleport_t` function.  If the layout is `Distillery` then when the teleportation replaces operations with primitives it uses the `Distil` gate instead of `Cultivate`.

I added some files that hold the cirq circuits for different magic state distillation circuits, and their test file---which is empty right now.  I also gave the distil cost function its own file and test file which is also empty right now.

I have included new tests in the already existing test files.  I usually found places where `Cultivate` was being used in the test and added a version with `Distil` instead when it made sense.  So far the tests are passing.  I also tried the hello_estimate notebook with the `Distillery` layout and changing some parameters and it seems to be working.

In the architecture class, I added the `distillation_repetition` keyword which is how many times you want to recuse the distillation protocol, however, currently the estimate does not take into account the odds of success in successfully distilling a T state---that is whether all syndrome measurements are +1.  I'm not sure how to include that.

### CODE REVIEW

File: resource_estimation/lattice_surgery_primitives.py [hunk 1]
```diff
@@ -237,6 +238,37 @@ def _json_namespace_(cls) -> str:
...
+
+    Distil|0^16> --> (|0> + e^(iθ)|1>)/√2 |0^15>
+    
+    """
+
+    def __init__(self,):
+        pass
+    
+    def num_qubits(self,):
+        return 1
```
Comment (ColinCampbellCQ) [inline resource_estimation/lattice_surgery_primitives.py:255]: Distillation acts on 16 logical qubit patches
```suggestion
        return 16
```

File: resource_estimation/compile_ftqc.py [hunk 2]
```diff
@@ -105,6 +103,10 @@ def replace_cirq_op(
 
 
 def teleport_T(op: cirq.Operation, layout: Layout) -> list[cirq.Operation]:
+    if hasattr(layout, "distil"):
+        distil = True
+    else:
+        distil = False
```
Comment (ColinCampbellCQ) [inline resource_estimation/compile_ftqc.py]: Fewer lines
```suggestion
    distil = hasattr(layout, "distil")
```

File: resource_estimation/compile_ftqc.py [hunk 3]
```diff
@@ -113,7 +115,10 @@ def teleport_T(op: cirq.Operation, layout: Layout) -> list[cirq.Operation]:
     ]
     operations = []
     if not available_t_factories:
-        operations += [lsp.Cultivate(pi / 4).on(factory) for factory in all_t_factories]
+        if distil:
+            operations += [lsp.Distil().on(factory) for factory in all_t_factories]
```
Comment (ColinCampbellCQ) [inline resource_estimation/compile_ftqc.py:116]: I think that Distillation should be registered as an operation on multiple logical qubits, which will break down this logic. We probably need to alter the definition of a factory from a qubit to a group of qubits. Then it would make this something like
```
[lsp.Distil().on(*factory) for factory in all_t_factories]
```

File: resource_estimation/layout.py [hunk 4]
```diff
@@ -454,3 +455,161 @@ def _generate(self) -> None:
...
+    #    | T | D | T | D | T
+    #  a | a | a | a | a | a
+    #  a | D | a | D | a | D
+    #  . | . | . | . | . | .   
+
+    def __init__(
+            self,
+            input_circuit: cirq.Circuit,
+            num_t_factories: int = 0,
+            num_s_factories: int = 0,
```
Comment (ColinCampbellCQ) [inline resource_estimation/layout.py]: Probably want to default to at least 1
```suggestion
            num_t_factories: int = 1,
            num_s_factories: int = 1,
```

File: resource_estimation/magic_state_distillation.py [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from collections.abc import Sequence
+
+import cirq
+import numpy as np
+import sympy
+
+
+def msd_5_to_1() -> cirq.Circuit:
```
Comment (ColinCampbellCQ) [inline resource_estimation/magic_state_distillation.py:21]: Is this the same as the implementation in [client superstaq](https://github.com/Infleqtion/client-superstaq/blob/main/cirq-superstaq/cirq_superstaq/circuits/msd.py)? If so, we can import it directly.

**Review by ColinCampbellCQ — COMMENTED** [review]
This is a great start! Let's sync up this week and try to add some more detail to the layout structure and the resource costs!

File: resource_estimation/compile_ftqc.py (hunk 3 above)
Comment (junmuthy) [reply resource_estimation/compile_ftqc.py:116]: Yes, I agree.  This is why i made the gate act on 1 qubit since It would still work even if it wasn't technically true.

File: resource_estimation/magic_state_distillation.py (hunk 5 above)
Comment (junmuthy) [reply resource_estimation/magic_state_distillation.py:21]: It is almost the same, but the superstaq version does include the repeat until success part of the circuit.  I have removed that part for the resource estimation.

File: resource_estimation/lattice_surgery_primitives.py (hunk 1 above)
Comment (junmuthy) [reply resource_estimation/lattice_surgery_primitives.py:255]: I will fix this as I also fix the register that Distil acts on.
